### PR TITLE
Expose realm ref when using RealmProvider in @realm/react

### LINKED ISSUE #4571: Expose realm ref when using RealmProvider in @realm/react
### Problem

Hello,

Thank you for this amazing service. When opening a `Realm`, one could handle client resets with the following code:

```ts
const handleRealmSyncError: Realm.ErrorCallback = useCallback(
  (_session, error) => {
    if (realmRef.current) {
      if (error.name === 'ClientReset') {
        realmRef.current.close();
        Realm.App.Sync.initiateClientReset(realmApp, realmPath);
        realmRef.current = null;
      }
    }
  },
  [realmPath],
);
```

and opening a Realm with the following sync config:

```ts
sync: {
  user: currentRealmUser,
  error: handleRealmSyncError,
},
```

But with `@realm/react`, we cannot use the `useRealm` hook in the same component where `RealmProvider` is declared, and thus there is no way to close the `Realm` in the error handler.

This enhancement should be a high priority as the production app would simply break in the event of a client reset and users would have to reinstall a newer version of the app with the updated schema.

### Solution

As mentioned by @takameyer, we should provide a ref prop to the RealmProvider and set the value when it’s rendered, so that it can be accessed in the error handler in the same component.

### How important is this improvement for you?

Dealbreaker

### PR DESCRIPTION
**Pull request by takameyer**
## What, How & Why?
It was found that we were missing a critical feature to the `RealmProvider`, which was a way to access the configured realm from outside of the provider.  This is being done through a forwarded ref, which is passed as a property into the provider.
Upon implementation, it was also apparent that this would be a useful addition to `AppProvider` as well.  

This closes #4571

## ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [ ] 🔀 Executed flexible sync tests locally if modifying flexible sync
* [x] 📱 Check the React Native/other sample apps work if necessary
* [x] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [ ] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: packages/realm-react/README.md [hunk 1]
```diff
@@ -250,6 +250,20 @@ const AppWrapper = () => {
   )
 }
 ```
+
+In some cases, it may be necessary to access the configured Realm from outside of the `RealmProvider`, for instance, implementing a client reset fallback.  This can be done by creating a `ref` with `useRef` and setting the `realmRef` property of `RealmProvider`.
+
+```tsx
+const AppWrapper = () => {
+  const realmRef = useRef<Realm|null>(null)
+
```
Comment (tomduncalf) [inline packages/realm-react/README.md:259]: Would it be worth showing a simple config here that uses the `realmRef`, just so it's really clear how to use this?

```suggestion
// Handle client reset errors by closing the Realm and initiating a client reset
const handleRealmSyncError: Realm.ErrorCallback = useCallback(
  (_session, error) => {
    if (realmRef.current) {
      if (error.name === 'ClientReset') {
        realmRef.current.close();
        Realm.App.Sync.initiateClientReset(realmApp, realmPath);
      }
    }
  },
  [realmPath],
);
```

Comment (tomduncalf) [reply packages/realm-react/README.md:259]: Maybe there is something even simpler we could show though...

**Review by tomduncalf — APPROVED** [review]
Nice solution 👍

**Comment by kraenhansen** [comment]
An alternative solution to the original issue would be to pass the realm instance into the callback provided through `config.sync.error`. Right now it takes `session` and `error`, how about exposing the `Realm` instance on the `session` object instead?

**Comment by tomduncalf** [comment]
I wondered about the same @kraenhansen but thought maybe this solution was more general, e.g. if you wanted to access the Realm in your top level component for some other reason. Maybe you're right that we should favour simplicity unless we can find another use case though

**Comment by takameyer** [comment]
>An alternative solution to the original issue would be to pass the realm instance into the callback provided through config.sync.error. Right now it takes session and error, how about exposing the Realm instance on the session object instead? That would be way simpler in my opinion.

@tomduncalf @kraenhansen 
In the linked issue, the user would need access to both the Realm and the Realm.App.  Would you also want to provide that app within the session object as well?  I'm not so familiar with the session object, but I'll take a look at the code. 

I thought this solution would be more flexible, since it would cover other use cases.  Adding realm to `session` would solve only the current issue.  On the other hand, maybe this is the only use case.  Thoughts?

File: packages/realm-react/README.md (hunk 1 above)
Comment (takameyer) [reply packages/realm-react/README.md:259]: I originally did throw that in there, which was what inspired me to add this strategy to the `AppProvider` as well.  I would like to review how a client reset is done and see if I can simplify this.

Comment (takameyer) [reply packages/realm-react/README.md:259]: I think I'll leave this as is for now.  I don't want to show deprecated use cases in the documentation.

**Comment by thespacemanatee** [comment]
@takameyer Is it possible to release a beta-6 for hermes builds and bump @realm/react to include this commit? Would be nice to have `ClientReset` functionality as we are gearing up for our first release soon, thank you!

**Comment by takameyer** [comment]
@thespacemanatee I'm releasing `@realm/react` v0.3.1 now.

**Comment by takameyer** [comment]
@thespacemanatee thanks for the heads up.  We will get that taken care of.

**Pull request merged**